cpp/include/raft/neighbors/ivf_flat_types.hpp

```diff
@@ -65,6 +77,37 @@ struct search_params : ann::search_params {
 static_assert(std::is_aggregate_v<index_params>);
 static_assert(std::is_aggregate_v<search_params>);
 
+/** The data for a single list (cluster). */
+template <typename T, typename IdxT, typename SizeT = uint32_t>
+struct list_data {
+  /** Cluster data. */
+  device_matrix<T, SizeT, row_major> data;
+  /** Source indices. */
+  device_vector<IdxT, SizeT> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
+
+  list_data(raft::device_resources const& res, SizeT n_rows, uint32_t dim)
+    : size{n_rows}
+  {
+    auto capacity = round_up_safe<SizeT>(bound_by_power_of_two<SizeT>(size), kIndexGroupSize);
+    try {
+      data = make_device_matrix<T, SizeT, row_major>(res, capacity, dim);
+      indices = make_device_vector<IdxT, SizeT>(res, capacity);
+    } catch (std::bad_alloc& e) {
+      RAFT_FAIL(
+        "ivf-flat: failed to allocate a big enough index list to hold all data "
+        "(requested size: %zu records, selected capacity: %zu records). "
+        "Allocator exception: %s",
+        size_t(size),
+        size_t(capacity),
+        e.what());
+    }
+    // Fill the index buffer with a pre-defined marker for easier debugging
+    matrix::fill(res, indices.view(), ivf_flat::kInvalidRecord<IdxT>);
```

Comment:
NB: I had to switch back from raft primitive to thrust here, because of compilation errors on the cython side (no dependency on cuda code allowed) https://github.com/rapidsai/raft/pull/1249/commits/adf7872d9997eef5745787a26f692405aac3bf0d

Reply:
NB 2: the ivf-pq PR merged and the list data is in a separate module now to avoid code duplication between ivf-flat and ivf-pq.